### PR TITLE
Refactor state latex drawer

### PR DESCRIPTION
**Pull request by frankharkins**
### Summary

Currently, we have two different algorithms to display amplitudes as human-readable latex strings. This PR adapts the functions in [qiskit/visualization/array.py](https://github.com/Qiskit/qiskit-terra/compare/main...frankharkins:latex-refactor?expand=1#diff-fcd8b8c48a0ab596f79a1b15f85e7e15db709aa850a1f5850c485489640a251e) so they work for both drawers, and deprecates the duplicate drawer.

### Details and comments

- I went with the sympy approach of the ket drawer over the home-rolled algorithm in the array drawer.
- As a side effect, the array drawer may behave slightly differently, I've mentioned this in the releasenotes. The `precision` default in the matrix drawer needed increasing to continue recognising common amplitudes. The ket drawer should produce latex that renders identically so hopefully minimal disruption there.
- I've also deprecated some public functions (`num_to_latex_ket`, `numbers_to_latex_terms`) as they're quite specific to the drawer and I don't think it falls under Qiskit's scope to support them. I can't find any uses of them in the wild.

This PR supersedes #8551

### CODE REVIEW

**Review by ikkoham — COMMENTED** [review]
Thank you. The code is more concise and the output is easier to read. I will leave some minor comments.
image

File: qiskit/visualization/state_visualization.py [hunk 1]
```diff
@@ -1233,69 +1234,34 @@ def num_to_latex_ket(raw_value: complex, first_term: bool, decimals: int = 10) -
...
     The first non-zero term is treated differently. For this term a leading + is suppressed.
-
     Args:
         numbers: List of numbers to format
         decimals: Number of decimal places to round to (default: 10).
     Returns:
         List of formatted terms
     """
+    warnings.warn(
+        "~qiskit.visualization.state_visualization.num_to_latex_terms "
```
[the remark this answers is not in the document]
Comment (frankharkins) [reply qiskit/visualization/state_visualization.py:1259]: I recommend dropping support for this feature. I can't find any uses of it, and it's very specific to this drawing algorithm. Would it be enough to point people to sympy instead?

```
~qiskit.visualization.state_visualization.num_to_latex_ket is deprecated as of 0.23.0 and will be removed no earlier than 3 months after the release. If you want similar functionality, see sympy's `nsimplify` and `latex` functions.
```

File: qiskit/visualization/array.py [hunk 2]
```diff
@@ -239,10 +185,12 @@ def array_to_latex(array, precision=5, prefix="", source=False, max_size=8):
     if array.ndim <= 2:
         if isinstance(max_size, int):
             max_size = (max_size, max_size)
-        outstr = _matrix_to_latex(array, precision=precision, prefix=prefix, max_size=max_size)
+        outstr = _matrix_to_latex(array, decimals=precision, prefix=prefix, max_size=max_size)
```
[the remark this answers is not in the document]
Comment (frankharkins) [reply qiskit/visualization/array.py:188]: Whoops, thank you

[the remark this answers is not in the document]
Comment (frankharkins) [reply qiskit/visualization/array.py]: Whoops, thanks

File: qiskit/visualization/state_visualization.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ikkoham) [reply qiskit/visualization/state_visualization.py]: OK! I agree with the dropping support.

**Review by ikkoham — DISMISSED** [review]
Thank you for addressing my review comments. LGTM

**Review by jakelishman — APPROVED** [review]
I reworded the release note a little, but otherwise this looks fine to me to merge.  It's good to consolidate all the code in one place for consistency, and if it reduces some of the spaghetti back-and-forth we'd grown over time, I'm all for it.  Thanks!

**Pull request merged**